### PR TITLE
KAFKA-16012: Ensure new leader information merged correctly with the current metadata

### PR DESCRIPTION
**Pull request by philipnee**
A small bug in KIP-951 that when we perform metadata update during the leader change, we replace all topic metadata with the partial response from the fetch response, which leads to the consumer only fetching from a subset of partitions.

In this PR, the predicate we pass in to `MetadataCache#mergeWith` always return true, because we always want to retain the current topic metadata.

### CODE REVIEW

File: clients/src/test/java/org/apache/kafka/clients/MetadataTest.java [hunk 1]
```diff
@@ -1071,18 +1071,25 @@ protected boolean retainTopic(String topic, boolean isInternal, long nowMs) {
     public void testUpdatePartitionLeadership() {
         Time time = new MockTime();
```
Comment (philipnee) [inline clients/src/test/java/org/apache/kafka/clients/MetadataTest.java:1147]: We probably need to refactor this test a certain point.  I tried to do it but as I moved something, we immediately exceeded the 160 line limit.

**Comment by philipnee** [comment]
Here is the test results
```
================================================================================
SESSION REPORT (ALL TESTS)
ducktape version: 0.8.14
session_id:       2023-12-15--001
run time:         4 minutes 22.315 seconds
tests run:        4
passed:           4
flaky:            0
failed:           0
ignored:          0
================================================================================
test_id:    kafkatest.tests.core.group_mode_transactions_test.GroupModeTransactionsTest.test_transactions.failure_mode=clean_bounce.bounce_target=brokers
status:     PASS
run time:   1 minute 40.624 seconds
--------------------------------------------------------------------------------
test_id:    kafkatest.tests.core.group_mode_transactions_test.GroupModeTransactionsTest.test_transactions.failure_mode=clean_bounce.bounce_target=clients
status:     PASS
run time:   2 minutes 5.353 seconds
--------------------------------------------------------------------------------
test_id:    kafkatest.tests.core.group_mode_transactions_test.GroupModeTransactionsTest.test_transactions.failure_mode=hard_bounce.bounce_target=clients
status:     PASS
run time:   2 minutes 3.331 seconds
--------------------------------------------------------------------------------
test_id:    kafkatest.tests.core.group_mode_transactions_test.GroupModeTransactionsTest.test_transactions.failure_mode=hard_bounce.bounce_target=brokers
status:     PASS
run time:   2 minutes 41.633 seconds
--------------------------------------------------------------------------------
```

File: clients/src/main/java/org/apache/kafka/clients/Metadata.java [hunk 2]
```diff
@@ -427,10 +427,15 @@ public synchronized Set<TopicPartition> updatePartitionLeadership(Map<TopicParti
             );
         }
 
-        this.cache = cache.mergeWith(cache.clusterResource().clusterId(), newNodes,
-                    updatePartitionMetadata, Collections.emptySet(), Collections.emptySet(),
-                    Collections.emptySet(), cache.cluster().controller(), topicIdsForUpdatedTopics,
-                    (topic, isInternal) -> !updatedTopics.contains(topic));
+        // Always merge the updated partitions with the existing topic metadata
```
Comment (jolshan) [inline clients/src/main/java/org/apache/kafka/clients/Metadata.java]: nit: Can we maybe include a note about the desired behavior -- FETCH requests can include some partition changes, and when we keep existing topic metadata, we take the new partition changes but keep all partitions that weren't changed?

File: clients/src/test/java/org/apache/kafka/clients/MetadataTest.java (hunk 1 above)
Comment (jolshan) [reply clients/src/test/java/org/apache/kafka/clients/MetadataTest.java:1147]: Maybe we should split up the test for each of the features (2 partitions, invalid topic, unauthorized topic, 1 internal topic). All the common logic can be in a helper?

Comment (philipnee) [reply clients/src/test/java/org/apache/kafka/clients/MetadataTest.java:1147]: sounds good - i was trying to limit the scope of change that's why i left this comment. i'll refactor this.

Comment (philipnee) [reply clients/src/test/java/org/apache/kafka/clients/MetadataTest.java:1147]: i did file this jira: https://issues.apache.org/jira/browse/KAFKA-16027

**Comment by philipnee** [comment]
hi @jolshan - thank you so much for the review. I have addressed the comment in the recent commits.

File: clients/src/test/java/org/apache/kafka/clients/MetadataTest.java [hunk 3]
```diff
@@ -1068,143 +1067,328 @@ protected boolean retainTopic(String topic, boolean isInternal, long nowMs) {
...
+        metadata.updateWithCurrentRequestVersion(
+            response,
+            false,
+            time.milliseconds());
+        assertEquals(2, metadata.fetch().partitionsForTopic(topic).size());
+        assertEquals(1, metadata.fetch().partition(tp0).leader().id());
+        assertEquals(1, metadata.fetch().partition(tp1).leader().id());
+
+        // "input-topic" partition 1 leader changes from node 1 to node 2
+        metadata.updatePartitionLeadership(
```
Comment (jolshan) [inline clients/src/test/java/org/apache/kafka/clients/MetadataTest.java:1131]: Is this the KIP-951 specific logic?

File: clients/src/test/java/org/apache/kafka/clients/MetadataTest.java [hunk 4]
```diff
@@ -1068,143 +1067,328 @@ protected boolean retainTopic(String topic, boolean isInternal, long nowMs) {
...
+        assertEquals(2, metadata.fetch().partition(tp1).leader().id());
+    }
+
+    @Test
+    public void testValidPartitionLeaderUpdate() {
         Time time = new MockTime();
 
-        // Setup metadata with initial set of 2 partitions, 1 each across topics, with 5 nodes.
-        // Also setup, 1 invalid topic, 1 unauthorized topic, 1 internal topic.
+        String clusterId = "kakfa-cluster";
```
Comment (jolshan) [inline clients/src/test/java/org/apache/kafka/clients/MetadataTest.java]: nit: kafka

File: clients/src/test/java/org/apache/kafka/clients/MetadataTest.java [hunk 5]
```diff
@@ -1068,143 +1067,328 @@ protected boolean retainTopic(String topic, boolean isInternal, long nowMs) {
...
+        // Since cluster metadata is unchanged, listener shouldn't be called.
+        verify(mockListener).onUpdate(any());
+        Mockito.reset(mockListener);
+    }
+    @Test
+    public void testInvalidPartitionLeaderUpdate() {
+        Time time = new MockTime();
+
+        String clusterId = "kakfa-cluster";
+        String unauthroizedTopics = "unauthorized-topic";
```
Comment (jolshan) [inline clients/src/test/java/org/apache/kafka/clients/MetadataTest.java]: nit: unauthorized and kafka

File: clients/src/test/java/org/apache/kafka/clients/MetadataTest.java (hunk 1 above)
Comment (jolshan) [reply clients/src/test/java/org/apache/kafka/clients/MetadataTest.java:1147]: Sorry for confusion, I think you were right to limit the scope of the change. Let's do the refactor as a followup.

File: clients/src/test/java/org/apache/kafka/clients/MetadataTest.java (hunk 3 above)
Comment (philipnee) [reply clients/src/test/java/org/apache/kafka/clients/MetadataTest.java:1131]: yes this was introduced in kip-951

File: clients/src/test/java/org/apache/kafka/clients/MetadataTest.java (hunk 1 above)
Comment (philipnee) [reply clients/src/test/java/org/apache/kafka/clients/MetadataTest.java:1147]: Alright - I'll revert some commits and re-push

**Comment by philipnee** [comment]
Hey @jolshan - Thanks again for the review.  Addressed your comments in the latest commits. The only client-related failure is `testExpandingTopicSubscriptions`, a known flaky test.  I will disable it in a different PR.

**Comment by msn-tldr** [comment]
@philipnee lgtm, thanks for working on this.

File: clients/src/test/java/org/apache/kafka/clients/MetadataTest.java (hunk 3 above)
Comment (jolshan) [reply clients/src/test/java/org/apache/kafka/clients/MetadataTest.java:1131]: Maybe we could just have a comment there (in the 1130 comment) 
something like: In KIP-951 we can update leadership info for a single partition. "input-topic partition 1...(everything else you said)

File: clients/src/test/java/org/apache/kafka/clients/MetadataTest.java [hunk 6]
```diff
@@ -1067,25 +1067,102 @@ protected boolean retainTopic(String topic, boolean isInternal, long nowMs) {
...
+        // topic1 has 2 partitions: tp11, tp12
+        // topic2 has 1 partition: tp21
         String topic1 = "topic1";
-        TopicPartition partition1 = new TopicPartition(topic1, 0);
-        PartitionMetadata part1Metadata = new PartitionMetadata(Errors.NONE, partition1, Optional.of(1), Optional.of(100), Arrays.asList(1, 2), Arrays.asList(1, 2), Arrays.asList(3));
+        TopicPartition tp11 = new TopicPartition(topic1, 0);
+        PartitionMetadata part1Metadata = new PartitionMetadata(Errors.NONE, tp11, Optional.of(1), Optional.of(100), Arrays.asList(1, 2), Arrays.asList(1, 2), Arrays.asList(3));
         Uuid topic1Id = Uuid.randomUuid();
+        TopicPartition tp12 = new TopicPartition(topic1, 1);
+        PartitionMetadata part12Metadata = new PartitionMetadata(Errors.NONE, tp12, Optional.of(2), Optional.of(200), Arrays.asList(2, 3), Arrays.asList(2, 3), Arrays.asList(1));
```
Comment (jolshan) [inline clients/src/test/java/org/apache/kafka/clients/MetadataTest.java:1161]: nit: (we can address in a followup refactor) this naming is a bit confusing. Maybe all should be part11Metadata, part12Metadata, part21Metadata or even just tp10Metadata, tp11Metadata etc.

File: clients/src/test/java/org/apache/kafka/clients/MetadataTest.java [hunk 7]
```diff
@@ -1185,9 +1263,14 @@ else if (partition.equals(internalPart))
     /**
      * For testUpdatePartially, validates that updatedMetadata is matching expected part1Metadata, part2Metadata, interalPartMetadata, nodes & more.
      */
-    void validateForUpdatePartitionLeadership(Metadata updatedMetadata, PartitionMetadata part1Metadata, PartitionMetadata part2Metadata, PartitionMetadata internalPartMetadata,
-        List<Node> expectedNodes, String expectedClusterId, Set<String> expectedUnauthorisedTopics, Set<String> expectedInvalidTopics, Set<String> expectedInternalTopics,
-        Node expectedController, Map<String, Uuid> expectedTopicIds) {
+    void validateForUpdatePartitionLeadership(Metadata updatedMetadata,
+                                              PartitionMetadata part1Metadata, PartitionMetadata part2Metadata, PartitionMetadata part12Metadata,
```
Comment (jolshan) [inline clients/src/test/java/org/apache/kafka/clients/MetadataTest.java:1267]: nit: (for refactor) -- if we decided to put all the arguments on new lines, let's do it for the all the arguments. The ones with three are a bit harder to read.

**Comment by jolshan** [comment]
I was looking at some of the test failures and wasn't seeing the OffsetsApiIntegrationTests failing elsewhere. Can we check those? Running locally may not be enough so I will also restart the build.

File: clients/src/test/java/org/apache/kafka/clients/MetadataTest.java (hunk 6 above)
Comment (philipnee) [reply clients/src/test/java/org/apache/kafka/clients/MetadataTest.java:1161]: yap - i was trying to be consistent with the previous variable names such as `PartitionMetadata part1Metadata`. I'll follow up with the changes in a separated refactor PR.

File: clients/src/test/java/org/apache/kafka/clients/MetadataTest.java (hunk 7 above)
Comment (philipnee) [reply clients/src/test/java/org/apache/kafka/clients/MetadataTest.java:1267]: thanks - would you mind leaving this for the refactor PR?

Comment (jolshan) [reply clients/src/test/java/org/apache/kafka/clients/MetadataTest.java:1267]: yup that's fine.

**Comment by philipnee** [comment]
hi @jolshan - it seems like `OffsetsApiIntegrationTests` is connect related test, in `org.apache.kafka.connect.integration`.  in the other PR, some of the tests in the integration test module also fail sometimes.

**Comment by jolshan** [comment]
I'm seeing some OOMs and such again. I will run one more time. @philipnee if you don't mind just double checking locally for the tests on your end. I'm not seeing any failing consistently, but just want to be extra careful near the code freeze. I will check back in few hours or so.

**Comment by philipnee** [comment]
Hey @jolshan - Thanks.  OOM also happens to another PR of mine see #15035 - i thought it was introduced by the AsyncKafkaConsumerTest but apparently there's more to it... Let me verify the tests locally and I'll the results there.

**Comment by philipnee** [comment]
image

Unit tests seems fine, there is 1 known flaky test under AbstractCoordinator and 1 related to AsyncKafkaConsumer.  Both are unrelated to the metadata change here.

**Comment by philipnee** [comment]
Consumer Integration tests also seems fine.
image

**Comment by jolshan** [comment]
Thanks for checking @philipnee. I will wait for this run to complete and we should be good to go.

**Review by jolshan — APPROVED** [review]

**Comment by philipnee** [comment]
Thank you!

**Comment by philipnee** [comment]
jenkins doesn't seem to be in a great state.  there are a bunch of zkMigrationIntegrationTest and ReplicaManagerTest complaining timeout and `Found unexpected 1 NonDaemon threads` (there are others as well)

**Comment by jolshan** [comment]
Yeah. I was looking at builds with @dajac for his PR too. The java 8 one here OOM'd. 
I think the ZkMigrationTests are causing the thread leak for the replica manager since I see replica manager tests failing with this issue on the same builds as the ZKMigrationTest failures and not a lot of commonalities otherwise.

I will file a JIRA for that. I will start one more time to get a super clear signal and will come back afterwards.

**Comment by jolshan** [comment]
jira here: [KAFKA-16045](https://issues.apache.org/jira/browse/KAFKA-16045)

**Comment by philipnee** [comment]
Hey @jolshan seems like the build finished without oom, what do you think?

**Comment by jolshan** [comment]
Ok -- looks good. I will merge and backport to 3.7.

**Pull request merged**